stan/math/rev/functor/algebra_solver_newton.hpp

```diff
@@ -54,22 +54,24 @@ namespace math {
  * @throw <code>std::invalid_argument</code> if max_num_steps is not positive.
  * @throw <code>std::domain_error</code> if solver exceeds max_num_steps.
  */
-template <typename F, typename T>
+template <typename F, typename T, require_eigen_vector_t<T>* = nullptr>
 Eigen::VectorXd algebra_solver_newton(
-    const F& f, const Eigen::Matrix<T, Eigen::Dynamic, 1>& x,
-    const Eigen::VectorXd& y, const std::vector<double>& dat,
-    const std::vector<int>& dat_int, std::ostream* msgs = nullptr,
-    double scaling_step_size = 1e-3, double function_tolerance = 1e-6,
+    const F& f, const T& x, const Eigen::VectorXd& y,
+    const std::vector<double>& dat, const std::vector<int>& dat_int,
+    std::ostream* msgs = nullptr, double scaling_step_size = 1e-3,
+    double function_tolerance = 1e-6,
```

Comment:
template for `y`'s type as well. Also the row/col vector thing

Reply:
Oh I'm seeing now we have two overloads for this function. @bbbales2 why is that?

Reply:
Looks like the first version is basically just a double version. `x` (initial guess) can be vars there but the var aspects will get ignored (this assumes the solution, if it exists, does not depend on the initial conditions which makes sense for an optimization).

Reply:
Templating y is not actually necessary, because the scalar is not templated. The compiler will know that it needs to convert y, so this is not a problem for accepting expressions. 

We still could do it, it it turns out we don't need it to be a plain type. However I tried not to dig too deep and waste time in those more complicated higher order functions.

Reply:
> But after a quick look in the guts of the solvers I think we do need x and y to be plain types. That is also the reason for the evals in next questions.

Hmm yeah I'd prefer the plain type here if you don't mind.

Reply:
I dont understant. `Eigen::VectorXd` is a plain type.